opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py

```diff
@@ -12,7 +12,214 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+
+from collections import OrderedDict
+from collections import deque
+from collections import namedtuple
+import threading
+import typing
+
 from opentelemetry import trace as trace_api
+from opentelemetry.sdk import util
+
+try:
+    from collections.abc import MutableMapping
+    from collections.abc import Sequence
+except ImportError:
+    from collections import MutableMapping
+    from collections import Sequence
+
+
+MAX_NUM_ATTRIBUTES = 32
+MAX_NUM_ANNOTATIONS = 32
+MAX_NUM_EVENTS = 128
+MAX_NUM_LINKS = 32
+
+AttributeValue = typing.Union[str, bool, float]
+
+
+class BoundedList(Sequence):
+    """An append only list with a fixed max size."""
+    def __init__(self, maxlen):
+        self.dropped = 0
+        self._dq = deque(maxlen=maxlen)
+        self._lock = threading.Lock()
+
+    def __repr__(self):
+        return ("{}({}, maxlen={})"
+                .format(
+                    type(self).__name__,
+                    list(self._dq),
+                    self._dq.maxlen
+                ))
+
+    def __getitem__(self, index):
+        return self._dq[index]
+
+    def __len__(self):
+        return len(self._dq)
+
+    def __iter__(self):
+        return iter(self._dq)
```

Comment:
Do we intend to handle `RuntimeError: deque mutated during iteration` or leave it as caller responsibility?

Reply:
How about making a defensive copy? 7836995

Reply:
How about making a defensive copy? 7836995

Reply:
Given most of the built-in collections throw exception when changes happened while iterating, it is fine to leave it. Unless we want this `__iter__` to be bulletproof. I asked this question just to confirm this is what we intend.

BTW, what is `7836995`?

Reply:
https://github.com/open-telemetry/opentelemetry-python/commit/7836995b4fde6cb5ff8634675d3e24415fa93372

Catching this in the exporter (where we'd iterate through these lists) would be odd since the spans should be finished at that point, but I think it's better to make these bulletproof than risk the unhandled exception later.